File: src/backend/go/internal/dto/todo_item.go
package dto

import "time"

// CreateTodoItemRequest mirrors a CreateTodoItemRequest command DTO in C#.
// The `binding` tag drives Gin's validation (analogous to [Required] / FluentValidation).
type CreateTodoItemRequest struct {
	Title       string  `json:"title"       binding:"required,min=1,max=200"`
	Description *string `json:"description" binding:"omitempty,max=2000"`
}

// UpdateTodoItemRequest mirrors an UpdateTodoItemRequest DTO — all fields optional.
type UpdateTodoItemRequest struct {
	Title       *string `json:"title"       binding:"omitempty,min=1,max=200"`
	Description *string `json:"description" binding:"omitempty,max=2000"`
	IsCompleted *bool   `json:"isCompleted"`
}

// TodoItemResponse mirrors a TodoItemDto / view model returned from controllers.
type TodoItemResponse struct {
	ID          uint       `json:"id"`
	Title       string     `json:"title"`
	Description *string    `json:"description"`
	IsCompleted bool       `json:"isCompleted"`
	CreatedAt   time.Time  `json:"createdAt"`
	UpdatedAt   *time.Time `json:"updatedAt"`
}

// PaginatedResponse is a generic paginated list wrapper — mirrors PagedResult<T> in C#.
type PaginatedResponse[T any] struct {
	Items      []T   `json:"items"`
	Total      int64 `json:"total"`
	Page       int   `json:"page"`
	PageSize   int   `json:"pageSize"`
	TotalPages int   `json:"totalPages"`
}

File: src/backend/python/api/schemas/todo_item.py
from datetime import datetime

from pydantic import BaseModel, Field


# ── Create DTO (analogous to a CreateTodoItemRequest / command model) ──────────

class CreateTodoItemRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=200, examples=["Buy groceries"])
    description: str | None = Field(default=None, max_length=2000, examples=["Milk, eggs, bread"])


# ── Update DTO ─────────────────────────────────────────────────────────────────

class UpdateTodoItemRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200, examples=["Buy groceries"])
    description: str | None = Field(default=None, max_length=2000, examples=["Milk, eggs, bread"])
    is_completed: bool | None = Field(default=None)


# ── Response DTO (analogous to a TodoItemDto / view model) ────────────────────

class TodoItemResponse(BaseModel):
    model_config = {"from_attributes": True}  # like AutoMapper / ORM-mapped DTO

    id: int
    title: str
    description: str | None
    is_completed: bool
    created_at: datetime
    created_by_user_id: int | None = None
    updated_at: datetime | None
    updated_by_user_id: int | None = None


# ── Paginated list response ────────────────────────────────────────────────────

class PaginatedResponse[T](BaseModel):
    items: list[T]
    total: int
    page: int
    page_size: int
    total_pages: int


# ── CSV import/export DTOs ─────────────────────────────────────────────────────

class ImportRowError(BaseModel):
    row: int
    error: str


class ImportResult(BaseModel):
    imported: int
    failed: int
    errors: list[ImportRowError] = Field(default_factory=list)
